coderd/database/queries/insights.sql

```diff
@@ -771,3 +771,66 @@ SELECT
 FROM unique_template_params utp
 JOIN workspace_build_parameters wbp ON (utp.workspace_build_ids @> ARRAY[wbp.workspace_build_id] AND utp.name = wbp.name)
 GROUP BY utp.num, utp.template_ids, utp.name, utp.type, utp.display_name, utp.description, utp.options, wbp.value;
+
+-- name: GetUserStatusCountsByDay :many
+WITH dates AS (
+    -- Generate a series of dates between start and end
+    SELECT
+        day::date
+    FROM
+        generate_series(
+            date_trunc('day', @start_time::timestamptz),
+            date_trunc('day', @end_time::timestamptz),
+            '1 day'::interval
+        ) AS day
+),
+all_users AS (
+    -- Get all users who have had any status changes in or before the range
+    SELECT DISTINCT user_id
+    FROM user_status_changes
+    WHERE changed_at <= @end_time::timestamptz
+),
+initial_statuses AS (
+    -- Get the status of each user right before each day
+    SELECT DISTINCT ON (user_id, day)
+        user_id,
+        day,
+        new_status as status
+    FROM
+        all_users
+    CROSS JOIN
+        dates
+    LEFT JOIN LATERAL (
+        SELECT new_status, changed_at
+        FROM user_status_changes
+        WHERE user_status_changes.user_id = all_users.user_id
+        AND changed_at < day + interval '1 day'
+        ORDER BY changed_at DESC
+        LIMIT 1
+    ) changes ON true
+    WHERE changes.new_status IS NOT NULL
+),
+daily_status AS (
+    SELECT
+        d.day,
+        i.status,
+        i.user_id
+    FROM
+        dates d
+    LEFT JOIN
+        initial_statuses i ON i.day = d.day
+)
+SELECT
+    day,
+    status,
+    COUNT(*) AS count
+FROM
+    daily_status
+WHERE
+    status IS NOT NULL
+GROUP BY
+    day,
+    status
+ORDER BY
+    day ASC,
+    status ASC;
```

Comment:
What's the purpose of this query? Is it related to license utilization, and if so, do we need to encode knowledge in the query about transitions from utilization -> non and also the total utilization count outside the selected time range?

For arguments sake, let's say a user was active in the distant past but has been flip-flopping between dormant and suspended daily within and before the time range, what data will this query produce and will it be useful?

Reply:
It is intended to return a count of active, dormant and suspended users per day for the requested time period. So in your example, it will return something like:
|2024-12-01|active|0|
|2024-12-01|dormant|1|
|2024-12-01|suspended|0|
|2024-12-02|active|0|
|2024-12-02|dormant|0|
|2024-12-02|suspended|1|
|2024-12-03|active|0|
|2024-12-03|dormant|1|
|2024-12-03|suspended|0|
|2024-12-04|active|0|
|2024-12-04|dormant|0|
|2024-12-04|suspended|1|

The query itself does not at this stage know which of these status states consume a license and which don't. the plan is to graph all of these states and state/label that only active users consume license seats.